clang/lib/Sema/SemaBoundsSafety.cpp

```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
   return false;
 }
 
+SourceRange Sema::BoundsSafetySourceRangeFor(const CountAttributedType *CATy) {
```

Comment:
This entire function seems like a very bad idea to me: I don’t think Sema should ever have to fiddle with source locations in terms of character offsets or anything like that. This seems extremely brittle in case a user writes something like e.g. `__counted_by (  count  )` or uses macros (which you mention in a comment doesn’t work).

I think you should be able to get a hold of the corresponding `CountAttributedTypeLoc` somehow (e.g. via the `FieldDecl`, which has a `TypeSourceInfo*`, which you should be able to get the type loc from; you might have to do this a few levels up the call stack before we ever get here) and then use its source location and range.

Reply:
Yeah this isn't great. The whole function is essentially a workaround for the fact we don't have a good way to get `CountAttributedTypeLoc` from a `CountAttributedType`.

Given that I don't want to solve this problem right now I think simplest thing to do is just use the SourceRange from the count expression (which is the fallback for this function anyway). It does mean the SourceRange is slightly wrong (doesn't actually include the attribute text) but this isn't a deal breaker.

I can then tackle fixing this as part of the work  to fix the issue mentioned in this comment: https://github.com/llvm/llvm-project/pull/106321#discussion_r1735207309

Does that sound reasonable to you?

Reply:
I agree with Sirraide here, but would have to spend some time contemplating the implications of your proposal.

Reply:
I've dropped the hacky implementation in favor of just using the SourceRange attached to the count expression. I've filed #113582 to track resolving this properly.